File: src/main/java/com/chatsummary/bot/service/TelegramDownloadService.java
package com.chatsummary.bot.service;

import com.chatsummary.bot.model.ChatAttachment;
import com.chatsummary.bot.util.ImageDownscaler;
import java.io.IOException;
import org.bson.types.Binary;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.telegram.telegrambots.client.okhttp.OkHttpTelegramClient;
import org.telegram.telegrambots.meta.api.methods.GetFile;
import org.telegram.telegrambots.meta.exceptions.TelegramApiException;

@Service
public class TelegramDownloadService {

    private final OkHttpTelegramClient telegramClient;
    private final int imageMaxDimension;

    public TelegramDownloadService(
            @Value("${telegram.bot.token}") String botToken,
            @Value("${image.max-dimension:1024}") int imageMaxDimension
    ) {
        this.telegramClient = new OkHttpTelegramClient(botToken);
        this.imageMaxDimension = imageMaxDimension;
    }

    public ChatAttachment downloadPhoto(String fileId) {
        return download(fileId, null);
    }

    public byte[] downloadBytes(String fileId) {
        // Raw bytes; used for voice messages and video notes, which VoiceStorageService persists to disk.
        try {
            var file = telegramClient.execute(new GetFile(fileId));
            if (file.getFilePath() == null) {
                throw new IllegalStateException("Telegram didn't return file path");
            }
            try (var inputStream = telegramClient.downloadFileAsStream(file)) {
                return inputStream.readAllBytes();
            }
        } catch (TelegramApiException | IOException exception) {
            throw new IllegalStateException("Failed to download Telegram file", exception);
        }
    }

    private ChatAttachment download(String fileId, String contentType) {
        try {
            var file = telegramClient.execute(new GetFile(fileId));
            var filePath = file.getFilePath();
            if (filePath == null) {
                throw new IllegalStateException("Telegram didn't return file path");
            }

            try (var inputStream = telegramClient.downloadFileAsStream(file)) {
                var fileBytes = inputStream.readAllBytes();
                var resolvedType = contentType != null
                        ? contentType
                        : (filePath.toLowerCase().endsWith(".png") ? "image/png" : "image/jpeg");
                var downscaled = ImageDownscaler.downscale(fileBytes, resolvedType, imageMaxDimension);
                return new ChatAttachment(resolvedType, new Binary(downscaled));
            }
        } catch (TelegramApiException | IOException exception) {
            throw new IllegalStateException("Failed to download Telegram file", exception);
        }
    }
}

File: src/main/kotlin/com/chatsummary/bot/service/TelegramDownloadService.kt
package com.chatsummary.bot.service

import com.chatsummary.bot.model.ChatAttachment
import org.bson.types.Binary
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import org.telegram.telegrambots.client.okhttp.OkHttpTelegramClient
import org.telegram.telegrambots.meta.api.methods.GetFile
import java.io.InputStream

@Service
class TelegramDownloadService (
    @param:Value("\${telegram.bot.token}") private val botToken: String,
    ) {

    private val telegramClient = OkHttpTelegramClient(botToken)

    fun downloadPhoto(fileId: String) : ChatAttachment {

        val getFileMethod = GetFile(fileId)

        // 2. Выполняем запрос к API Telegram для получения метаданных файла (включая путь)

        val fileObj = telegramClient.execute(getFileMethod)
        val filePath = fileObj.filePath
            ?: throw IllegalStateException("Telegram не вернул путь к файлу")

        // 3. Используем встроенный в SDK метод для скачивания файла напрямую в InputStream
        val inputStream: InputStream = telegramClient.downloadFileAsStream(fileObj)

        // 4. Вычитываем поток в массив байт
        val fileBytes = inputStream.use { it.readBytes() }

        // Определяем contentType по расширению
        val contentType = if (filePath.endsWith(".png", true)) "image/png" else "image/jpeg"

        return ChatAttachment(
            contentType = contentType,
            data = Binary(fileBytes)
        )
    }

}
